File: 01_linear_probing.py
class LinearProbingHashTable:
    def __init__(self, size):
        self.size = size
        self.table = [None] * size

    def hash_function(self, key):
        return key % self.size

    def insert(self, key):
        idx = self.hash_function(key)
        start = idx
        while self.table[idx] is not None:
            idx = (idx + 1) % self.size
            if idx == start:
                raise Exception("Hash Table is full")
        self.table[idx] = key

    def search(self, key):
        idx = self.hash_function(key)
        start = idx
        while self.table[idx] is not None:
            if self.table[idx] == key:
                return True
            idx = (idx + 1) % self.size
            if idx == start:
                break
        return False

if __name__ == "__main__":
    h = LinearProbingHashTable(7)
    for k in [10, 20, 5, 15, 7]:
        h.insert(k)
    print("Table:", h.table)
    print("Search 15:", h.search(15))
    print("Search 99:", h.search(99))

File: 01_linear_probing.c
#include <stdio.h>
#include <stdlib.h>

#define SIZE 7

int hashTable[SIZE];

void init() {
    for (int i = 0; i < SIZE; i++)
        hashTable[i] = -1;
}

int hashFunction(int key) {
    return key % SIZE;
}

void insert(int key) {
    int idx = hashFunction(key);
    int start = idx;
    while (hashTable[idx] != -1) {
        idx = (idx + 1) % SIZE;
        if (idx == start) {
            printf("Hash Table is full\n");
            return;
        }
    }
    hashTable[idx] = key;
}

int search(int key) {
    int idx = hashFunction(key);
    int start = idx;
    while (hashTable[idx] != -1) {
        if (hashTable[idx] == key) return 1;
        idx = (idx + 1) % SIZE;
        if (idx == start) break;
    }
    return 0;
}

int main() {
    init();
    int keys[] = {10, 20, 5, 15, 7};
    for (int i = 0; i < 5; i++)
        insert(keys[i]);

    printf("Hash Table: ");
    for (int i = 0; i < SIZE; i++)
        printf("%d ", hashTable[i]);
    printf("\n");

    printf("Search 15: %d\n", search(15));
    printf("Search 99: %d\n", search(99));
    return 0;
}
